sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/cosmos/CosmosAutoConfiguration.java

```diff
@@ -27,44 +40,96 @@
 @ConditionalOnClass({ CosmosAsyncClient.class, CosmosTemplate.class })
 @ConditionalOnResource(resources = "classpath:cosmos.enable.config")
 @EnableConfigurationProperties(CosmosProperties.class)
+@AutoConfigureAfter(AzureDefaultTokenCredentialAutoConfiguration.class)
 public class CosmosAutoConfiguration extends AbstractCosmosConfiguration {
-    private final CosmosProperties cosmosProperties;
-    private final AzureProperties azureProperties;
-
+    private final CosmosProperties properties;
+    private static final String COSMOS_CHAINED_TOKEN_CREDENTIAL_BEAN_NAME = "cosmosChainedTokenCredential";
+    private static final String COSMOS_AZURE_KEY_CREDENTIAL_BEAN_NAME = "cosmosAzureKeyCredential";
 
-    public CosmosAutoConfiguration(CosmosProperties cosmosProperties,
-                                   @Qualifier(AZURE_PROPERTY_BEAN_NAME) AzureProperties azureProperties) {
-        this.cosmosProperties = cosmosProperties;
-        this.azureProperties = azureProperties;
+    public CosmosAutoConfiguration(CosmosProperties properties) {
+        this.properties = properties;
     }
 
     @Override
     protected String getDatabaseName() {
-        return cosmosProperties.getDatabase();
+        return properties.getDatabase();
+    }
+
+    @Bean(COSMOS_AZURE_KEY_CREDENTIAL_BEAN_NAME)
+    @ConditionalOnMissingBean(name = COSMOS_AZURE_KEY_CREDENTIAL_BEAN_NAME)
+    public AzureKeyCredential cosmosAzureKeyCredential() {
+        return Optional.ofNullable(properties.getKey())
+                       .filter(StringUtils::hasText)
+                       .map(AzureKeyCredential::new)
+                       .orElse(null);
+    }
+
+    @Bean(COSMOS_CHAINED_TOKEN_CREDENTIAL_BEAN_NAME)
+    @ConditionalOnMissingBean(name = COSMOS_CHAINED_TOKEN_CREDENTIAL_BEAN_NAME)
+    public ChainedTokenCredential cosmosChainedTokenCredential(
+        @Qualifier(DEFAULT_CHAINED_TOKEN_CREDENTIAL_BEAN_NAME) TokenCredential defaultTokenCredential) {
+        SpringMappingCredentialPropertiesProvider provider = new SpringMappingCredentialPropertiesProvider(properties);
+        final ChainedTokenCredentialBuilder chainedTokenCredentialBuilder = new ChainedTokenCredentialBuilder();
+        chainedTokenCredentialBuilder.addLast(provider.mappingTokenCredential())
+                                     .addLast(defaultTokenCredential);
+        return chainedTokenCredentialBuilder.build();
+    }
+
+    @Bean
+    @ConditionalOnMissingBean
+    public AzureKeyCredentialClientBuilderCustomizer<CosmosClientBuilder> azureKeyCredentialCustomizer(
+        @Autowired(required = false) @Qualifier(COSMOS_AZURE_KEY_CREDENTIAL_BEAN_NAME) AzureKeyCredential cosmosAzureKeyCredential) {
+        if (cosmosAzureKeyCredential != null) {
+            return builder -> builder.credential(cosmosAzureKeyCredential);
+        }
+        return null;
```

Comment:
we should discuss this design, I'm really not sure if it is the right way to do this

Reply:
btw, you can use `ObjectProvider` to get rid of this `@Autowired(required=false)`